### PR TITLE
Add Language.executable_extension property

### PR DESCRIPTION
**Pull request by andreyv**
This property allows to simplify language compilation commands in multiple cases. The compilation commands presented to the user are also easier to understand.

The initial motivation for this change was to be able to use [`python3 -m zipapp`](https://docs.python.org/3/library/zipapp.html) instead of manual archiving for the Python 3 language. Currently this doesn't work, however, because from [PEP 441](https://www.python.org/dev/peps/pep-0441/#a-new-python-zip-application-extension) there must be a `__main__.py` file in the archive, and `zipapp` [enforces](https://github.com/python/cpython/blob/b16e382c446d76ede22780b15c75f43c5f132e25/Lib/zipapp.py#L112-L117) this. But these changes are still useful for multiple other languages.

A couple of notes:
1. The `.pyz` extension is specified in [PEP 441](https://www.python.org/dev/peps/pep-0441/).
1. This code doesn't work if there are multiple source files, but no grader:
   https://github.com/cms-dev/cms/blob/d4c9e926bd52d8022069c417b206b0882ef4d1ba/cms/grading/tasktypes/Batch.py#L256-L257
    I think it could query the task object instead. Perhaps a new issue can be opened.
1. The PHP language plugin shows that it may be useful to have an `InterpretedLanguage` subclass without compilation commands. Currently a language with empty compilation commands fails here:
    https://github.com/cms-dev/cms/blob/d4c9e926bd52d8022069c417b206b0882ef4d1ba/cms/grading/steps/utils.py#L88-L89
    because `stats` is `None`. On the other hand, it is not clear how this can be reconciled with the case of multiple source files.
1. Logically, each language type plugin should implement its own updater function, so that the updater script can query all plugins, including out-of-tree ones. As this is overkill, the updater just has a hardcoded list of languages instead.

---
This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/1139)

### CODE REVIEW

File: cms/grading/languages/php.py [hunk 1]
```diff
@@ -40,11 +40,19 @@ def source_extensions(self):
...
+
     def get_compilation_commands(self,
                                  source_filenames, executable_filename,
                                  for_evaluation=True):
         """See Language.get_compilation_commands."""
-        return [["/bin/cp", source_filenames[0], executable_filename]]
+        if source_filenames[0] != executable_filename:
+            return [["/bin/cp", source_filenames[0], executable_filename]]
+        else:
+            return [["/bin/true"]]
```
Comment (stefano-maggiolo) [inline cms/grading/languages/php.py:55]: Can this be an empty list instead?

File: cms/grading/languages/python3_cpython.py [hunk 2]
```diff
@@ -45,30 +45,25 @@ def source_extensions(self):
         """See Language.source_extensions."""
         return [".py"]
 
+    @property
+    def executable_extension(self):
+        """See Language.executable.extension."""
+        return ".pyz"
```
Comment (stefano-maggiolo) [inline cms/grading/languages/python3_cpython.py:51]: Comment with a link to the PEP

File: cms/grading/languages/python3_cpython.py [hunk 3]
```diff
@@ -45,30 +45,25 @@ def source_extensions(self):
...
 
+        commands = [["/usr/bin/python3", "-m", "compileall", "-b", "."]]
+
+        # The file with the entry point must be in first position.
+        basename, _ = os.path.splitext(os.path.basename(source_filenames[0]))
+        pyc_filename = "%s.pyc" % basename
+        commands.append(["/bin/mv", pyc_filename, self.MAIN_FILENAME])
+
+        commands.append(["/usr/bin/zip", executable_filename,
+                         self.MAIN_FILENAME] + source_filenames[1:])
```
Comment (stefano-maggiolo) [inline cms/grading/languages/python3_cpython.py:66]: This zips the .py files, not the .pyc - is this intended?

File: cms/grading/languages/python2_cpython.py [hunk 4]
```diff
@@ -45,30 +45,25 @@ def source_extensions(self):
...
 
+        commands = [["/usr/bin/python2", "-m", "compileall", "."]]
+
+        # The file with the entry point must be in first position.
+        basename, _ = os.path.splitext(os.path.basename(source_filenames[0]))
+        pyc_filename = "%s.pyc" % basename
+        commands.append(["/bin/mv", pyc_filename, self.MAIN_FILENAME])
+
+        commands.append(["/usr/bin/zip", executable_filename,
+                         self.MAIN_FILENAME] + source_filenames[1:])
```
Comment (stefano-maggiolo) [inline cms/grading/languages/python2_cpython.py:66]: This zips the .py files, not the .pyc - is this intended?

File: cmscontrib/updaters/update_43.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+"""A class to update a dump created by CMS.
+
+Used by DumpImporter and DumpUpdater.
+
+This updater adds file name extensions to relevant executables.
+
+Note:
+  - File descriptions are not updated.
+  - Out-of-tree languages should be added to LANG_EXT manually.
```
Comment (stefano-maggiolo) [inline cmscontrib/updaters/update_43.py:27]: Add "- Custom use of USE_JAR = False should be reflected in this updater."

**Review by stefano-maggiolo — COMMENTED** [review]
What's the current status of TwoSteps? Are we letting it rot away?

File: cms/grading/languages/php.py (hunk 1 above)
Comment (andreyv) [reply cms/grading/languages/php.py:55]: There is a problem here: https://github.com/cms-dev/cms/blob/d4c9e926bd52d8022069c417b206b0882ef4d1ba/cms/grading/steps/utils.py#L88-L89

If there are no commands, then `stats` is `None`. I can see two possible ways: initialize `stats` with a "zero" object (or treat `None` as zero in relevant places), or skip compilation altogether if there are no commands. But the latter is likely to introduce special cases in multiple places.

This PR avoids the problem using `/bin/true` for now.

File: cms/grading/languages/python3_cpython.py (hunk 2 above)
Comment (andreyv) [reply cms/grading/languages/python3_cpython.py:52]: Added.

File: cms/grading/languages/python3_cpython.py (hunk 3 above)
Comment (andreyv) [reply cms/grading/languages/python3_cpython.py]: Nice catch! Fixed by returning the old implementation.

File: cms/grading/languages/python2_cpython.py (hunk 4 above)
Comment (andreyv) [reply cms/grading/languages/python2_cpython.py]: Fixed.

File: cmscontrib/updaters/update_43.py (hunk 5 above)
Comment (andreyv) [reply cmscontrib/updaters/update_43.py]: Added.

**Comment by andreyv** [comment]
> What's the current status of TwoSteps? Are we letting it rot away?

Thanks, I updated it as well (somewhat). Even in master, as I can see, TwoSteps won't work with Python etc., because it doesn't call `get_evaluation_commands()`, but rather just runs `./executable`. But at least now it is not broken more than it is in master.

File: cms/grading/languages/php.py (hunk 1 above)
Comment (stefano-maggiolo) [reply cms/grading/languages/php.py:55]: Ok, please add a comment saying why we need it.

**Comment by stefano-maggiolo** [comment]
I had one more comment request then feel free to push to master. Thank you!

File: cms/grading/languages/php.py (hunk 1 above)
Comment (andreyv) [reply cms/grading/languages/php.py:56]: Added a comment.

**Comment by andreyv** [comment]
Thanks! I'll push the changes soon.

**Comment by andreyv** [comment]
Merged.

**Pull request merged**